api/apps/api/test/access-control/project-acl.fixtures.ts

```diff
@@ -21,37 +22,95 @@ export const getFixtures = async () => {
   const projectViewerRole = Roles.project_viewer;
   const projectContributorRole = Roles.project_contributor;
   const projectOwnerRole = Roles.project_owner;
+  let projectCreatedId = '';
   const userProjectsRepo: Repository<UsersProjectsApiEntity> = app.get(
     getRepositoryToken(UsersProjectsApiEntity),
   );
   const cleanups: (() => Promise<void>)[] = [];
 
   return {
     cleanup: async () => {
-      await userProjectsRepo.delete({
-        userId: In([
-          ownerUserId,
-          contributorUserId,
-          viewerUserId,
-          otherOwnerUserId,
-        ]),
-      });
-      await Promise.all(cleanups.map((clean) => clean()));
+      for (const cleanup of cleanups.reverse()) {
+        await cleanup();
+      }
       await app.close();
     },
 
+    userRoleCleanup: async () => {
+      if (projectCreatedId) {
+        cleanups.push(
+          async () =>
+            await ProjectsACLTestUtils.deleteUserFromProject(
+              app,
+              ownerUserToken,
+              projectCreatedId,
+              otherOwnerUserId,
+            ),
+        );
+        cleanups.push(
+          async () =>
+            await ProjectsACLTestUtils.deleteUserFromProject(
+              app,
+              ownerUserToken,
+              projectCreatedId,
+              viewerUserId,
+            ),
+        );
+        cleanups.push(
+          async () =>
+            await ProjectsACLTestUtils.deleteUserFromProject(
+              app,
+              ownerUserToken,
+              projectCreatedId,
+              contributorUserId,
+            ),
+        );
+      }
+    },
+
+    GivenProjectExistsAndHasUsers: async () => {
+      const result = await userProjectsRepo.findOne({
+        where: {
+          userId: ownerUserId,
+          roleName: projectOwnerRole,
+        },
+      });
+
+      if (!result) {
+        throw new Error('Project does not exist');
+      }
+
+      return result.projectId;
+    },
```

Comment:
Seems like a "background" (see below) - and actually it is not creating/setting up anything  (`Given something happened`)

Reply:
I don't know if I get it completely but, I have two different things `GivenProjectWasCreated` and `GivenProjectExistsAndHasUsers`.

I'm using the first one for the addition of new users and to check who can or can't add any new users. I create this project from scratch because I can't use the ones already in the seeds.

The second one is used for the GET actions, as all the users are already part of the project, as well as for the DELETE of every type of user. This avoids not having to create a new project with every type of user for every test.

However, as I got from the Background link you posted, are those two not `Given` steps that repeat a lot but on different `Scenarios`? First method for when you need a project from scratch, the other already created projects with already established roles.

The only thing I realized is that by using second method for the DELETE of every type of user, I'm effectively not cleaning up/restoring the state of the seeds to its previous state, as the user are deleted from a specific project and not added again.

Reply:
as discussed:

```
  const { projectId, cleanup } = await GivenProjectExists(app, jwtToken);
  cleanups.push(cleanup)

//
const projectId = GivenProjectWasCreated()
const userA = GivenUserWasAddedToProject(projectId, Role)
const userB = GivenUserWasAddedToProject(projectId, Role)
const userC = GivenUserWasAddedToProject(projectId, Role)
```